# <font color=turquoise> Trabalhar com os dados variáveis em aplicativos de console C# 
___

## <font color=mediumorchid>  Escolher o tipo de dados correto em seu código C#
___

### <font color=hotpink> O que são dados? </font>

- No desenvolvimento de software, os dados são essencialmente um valor armazenado na memória do computador como uma série de bits.
- Um bit é uma opção binária simples representada como 0 ou 1, ou melhor, como "desligado" e "ligado"
- Quando usado em um byte, cada bit assume um significado na sequência. Na verdade, será possível representar 256 combinações diferentes com somente 8 bits se você usar um sistema numérico binário (base 2).
    - Por exemplo, em um sistema numérico binário, é possível representar o número 195 como 11000011. 

|128|64|32|16|8|4|2|1|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|1|1|0|0|0|0|1|1|

### <font color=hotpink> Tipos de valor versus tipos de referência </font>

- As variáveis de tipos de referência armazenam referências aos respectivos dados (objetos), ou seja, elas apontam para valores de dados armazenados em outro lugar. 
- As variáveis de tipos de valor contêm diretamente os respectivos dados. 

## <font color=mediumorchid>  Tipo Inteiro
___

Há duas subcategorias de tipos inteiros: tipos inteiros com sinal e sem sinal.

### <font color=hotpink> Usar as propriedades MinValue e MaxValue </font>

- Tipos com sinal:

In [2]:
Console.WriteLine("Valores dos tipos inteiros com sinal: \n");

Console.WriteLine($"sbyte:\t {sbyte.MinValue} até {sbyte.MaxValue}");
Console.WriteLine($"short:\t {short.MinValue} até {short.MaxValue}");
Console.WriteLine($"int:\t {int.MinValue} até {int.MaxValue}");
Console.WriteLine($"long:\t {long.MinValue} até {long.MaxValue}");

Valores dos tipos inteiros com sinal: 

sbyte:	 -128 até 127
short:	 -32768 até 32767
int:	 -2147483648 até 2147483647
long:	 -9223372036854775808 até 9223372036854775807


- Tipos sem sinal:

In [3]:
Console.WriteLine("Valores do tipo inteiro sem sinal: \n");

Console.WriteLine($"byte:\t {byte.MinValue} até {byte.MaxValue}");
Console.WriteLine($"ushort:\t {ushort.MinValue} até {ushort.MaxValue}");
Console.WriteLine($"uint:\t {uint.MinValue} até {uint.MaxValue}");
Console.WriteLine($"ulong:\t {ulong.MinValue} até {ulong.MaxValue}");


Valores do tipo inteiro sem sinal: 

byte:	 0 até 255
ushort:	 0 até 65535
uint:	 0 até 4294967295
ulong:	 0 até 18446744073709551615


## <font color=mediumorchid> Ponto flutuante
___

### <font color=hotpink> Tipos </font>

- Um ponto flutuante é um tipo de valor simples que representa números à direita da casa decimal. 
- Precisão é o número de casas com valores armazenadas após o ponto decimal.
- os valores `float` e `double` são armazenados internamente em um formato binário (base 2), enquanto `decimal` é armazenado em um formato decimal (base 10). 
- Geralmente, a matemática de ponto flutuante binária é uma aproximação do valor real. Isso significa que `float` e `double` são úteis, porque números grandes podem ser armazenados usando um pequeno volume de memória. 
- No entanto, `float` e `double` só devem ser usados quando uma aproximação é útil. Por exemplo, errar em alguns milhares ao calcular a zona de explosão de uma bola de neve em um videogame é uma aproximação boa o suficiente.
- Quando você precisar de uma resposta mais precisa, deverá usar `decimal`. Cada valor do tipo `decimal` tem um volume de memória relativamente grande; contudo, efetuar operações matemáticas dá a você um resultado mais preciso. Portanto, você deve usar o `decimal` ao trabalhar com os dados financeiros ou com qualquer cenário em que seja necessário um resultado preciso de um cálculo.

### <font color=hotpink> Usar as propriedades MinValue e MaxValue para cada tipo de ponto flutuante </font>

- Com sinal:

In [11]:
Console.WriteLine("Valores de ponto flutuante com sinal:\n");

Console.WriteLine($"float  :\t{float.MinValue} até {float.MaxValue}");
Console.WriteLine($"double :\t{double.MinValue} até {double.MaxValue}");
Console.WriteLine($"decimal:\t{decimal.MinValue} até {decimal.MaxValue}");


Valores de ponto flutuante com sinal:

float  :	-3,4028235E+38 até 3,4028235E+38
double :	-1,7976931348623157E+308 até 1,7976931348623157E+308
decimal:	-79228162514264337593543950335 até 79228162514264337593543950335


## <font color=mediumorchid> Tipos de Referência
___

- Uma variável de tipo de valor armazena os próprios valores diretamente em uma área de armazenamento chamada pilha. A pilha é a memória alocada ao código que está em execução, no momento, na CPU (também conhecida como o registro de ativação ou quadro de ativação). Quando a execução do registro de ativação for concluída, os valores na pilha serão removidos.

- Uma variável de tipo de referência armazena os próprios valores em uma região de memória separada chamada heap. O heap é uma área de memória compartilhada entre muitos aplicativos em execução no sistema operacional ao mesmo tempo. 
    - O Tempo de Execução do .NET comunica-se com o sistema operacional para determinar quais endereços de memória estão disponíveis e solicita um endereço em que ele possa armazenar o valor.
    - O runtime do .NET armazena o valor e, em seguida, retorna o endereço de memória à variável. 
    - Quando o código usa a variável, o Tempo de Execução do .NET pesquisa sem interrupção o endereço armazenado na variável e recupera o valor que está armazenado lá.

### <font color=hotpink> Definir uma variável de tipo de referência </font>

In [12]:
int[] data;

- Nesse momento, data é simplesmente uma variável que pode armazenar uma referência ou, em vez disso, um endereço de memória de um valor no heap. 

In [13]:
int[] data;

data = new int[3];

- A palavra-chave `new` informa ao runtime do `.NET` que ele deve criar uma instância da matriz `int` e, em seguida, coordenar-se com o sistema operacional para armazená-la na memória. O runtime do .NET cumpre e retorna um endereço de memória da nova matriz `int`. Finalmente, o endereço de memória é armazenado nos dados de variável. O padrão dos elementos da matriz `int` é o valor `0`, pois ele é o valor padrão de um `int`.

In [14]:
Console.WriteLine(data[1]);

0


- executar as duas operações em uma única linha de código:

In [15]:
int[] data = new int[3];

foreach (var i in data)  
{
    Console.WriteLine(i);
}

0
0
0


### <font color=hotpink>  Tipo de dados de cadeia de caracteres </font>

- O tipo de dados string também é um tipo de referência. 
- O operador new não foi usado ao declarar uma cadeia de caracteres. Isso é meramente uma conveniência proporcionada pelos designers do C#. Como o tipo de dados string é usado com tanta frequência, é possível usar o seguinte formato:

In [16]:
string shortenedString = "Hello World!";
Console.WriteLine(shortenedString);

Hello World!


## <font color=mediumorchid> Escolher o tipo de dados certo
___

### <font color=hotpink> Escolher o tipo de dados que atende aos requisitos de intervalo de valor de limite do seu aplicativo </font>

- Sua escolha de um tipo de dados pode ajudar a definir os limites do tamanho dos dados que você pode armazenar nessa variável específica. Por exemplo, se você sabe que uma variável específica deve armazenar apenas um número entre 1 e 10 mil, caso contrário, estaria fora dos limites do que seria esperado, você provavelmente evitará byte e sbyte porque seus intervalos são muito baixos.

- Além disso, você provavelmente não precisaria de int, long, uint e ulong, porque eles podem armazenar mais dados do que é necessário. Da mesma forma, você provavelmente ignoraria float, double e decimal se não precisasse de valores fracionários. Você provavelmente ficará em dúvida entre short e ushort, e ambos podem dar certo. Se você sabe que um valor negativo não tem sentido no aplicativo, pode escolher ushort (inteiro positivo sem sinal, de 0 a 65.535). Agora, qualquer valor atribuído a uma variável do tipo ushort fora do limite de 0 a 65535 geraria uma exceção, o que ajuda sutilmente a impor um grau de verificação de sanidade em seu aplicativo.

### <font color=hotpink> Comece escolhendo o tipo de dados correto para os dados (não para otimizar o desempenho) </font>

- Talvez você considere escolher o tipo de dados que usa o menor número de bits para o armazenamento de dados, achando que isso aprimorará o desempenho do aplicativo. No entanto, alguns dos melhores conselhos relacionados ao desempenho do aplicativo (ou seja, a velocidade de execução do seu aplicativo) é não "otimizar prematuramente". Não tente modificar as coisas nas partes do código, incluindo a seleção dos tipos de dados, que pode afetar o desempenho do aplicativo.

- Muitos pressupõem que, como um determinado tipo de dados armazena menos informações, ele deve usar menos o processador e a memória do computador do que um tipo de dados que armazena mais informações. Em vez disso, você deve escolher o ajuste correto para seus dados e, posteriormente, pode, de maneira empírica, medir o desempenho do seu aplicativo usando um software especial que fornece insights reais para as partes do seu aplicativo que afetam negativamente o desempenho.

### <font color=hotpink> Escolher tipos de dados com base nos tipos de dados de entrada e saída das funções de biblioteca usadas </font>

- Suponha que você queira trabalhar com um intervalo de anos entre duas datas. Como esse é um aplicativo de negócios, você pode determinar que só precisa de um intervalo entre cerca de 1960 a 2200. Isso pode fazer você pensar em utilizar byte, pois ele pode representar números entre 0 e 255.

- No entanto, quando você examina métodos internos nas classes System.TimeSpan e System.DateTime, percebe que elas aceitam valores do tipo double e int. Se escolher sbyte, você constantemente estará transmitindo para frente e para trás entre sbyte e double ou int. Nesse caso, poderá fazer mais sentido escolher int se você não precisar de uma precisão de subsegundos e double se você precisar de uma precisão de subsegundos.

### <font color=hotpink> Escolher tipos de dados com base no impacto em outros sistemas </font>

- Às vezes, é necessário considerar como as informações serão consumidas por outros aplicativos ou outros sistemas, como um banco de dados. Por exemplo, o sistema de tipo do SQL Server é diferente do sistema de tipo do C#. Como resultado, um mapeamento entre os dois deve acontecer antes que você possa salvar dados nesse banco de dados.

- Se a finalidade do aplicativo for a interação com um banco de dados, você provavelmente precisará considerar como esses dados são armazenados e em qual quantidade. A escolha de um tipo de dados maior pode afetar a quantidade (e o custo) de armazenamento físico necessário para armazenar todos os dados gerados pelo aplicativo.

### <font color=hotpink> Em caso de dúvida, vá pelo caminho mais simples </font>

Embora você tenha feito várias considerações, você está apenas começando. Por isso, para simplificar, você deve preferir um subconjunto de tipos de dados básicos, incluindo:

- `int` para a maioria dos números inteiros
- `decimal` para números que representam dinheiro
- `bool` para valores true ou false
- `string` para valor alfanumérico

### <font color=hotpink> Escolha tipos complexos especializados para situações especiais </font>

Não reinvente os tipos de dados se um ou mais já existirem para uma determinada finalidade. Os seguintes exemplos identificam quando um tipo de dados .NET específico pode ser útil:

- `byte`: para trabalhar com dados codificados provenientes de outros sistemas de computação ou que usam diferentes conjuntos de caracteres.
- `double`: para trabalhar com fins geométricos ou científicos. double é usado com frequência ao criar jogos que envolvem movimento.
- `System.DateTime` para um valor de data e hora específico.
- `System.TimeSpan` para um intervalo de anos/mês/dias/horas/minutos/segundos/milissegundos.

# <font color=turquoise> Converter tipos de dados usando técnicas de conversão cast em C#
___

## <font color=mediumorchid> Explorar a coerção e a conversão de tipo de dados
___

Existem várias técnicas para realizar a conversão de tipos de dados. A escolha da técnica adequada depende da resposta a duas questões cruciais:

1. Considerando o valor em questão, existe a possibilidade de que a tentativa de conversão de tipo de dados possa levar a uma exceção durante a execução do programa?
2. Dependendo do valor, é possível que a tentativa de conversão de tipo de dados resulte em perda de informações?


### <font color=hotpink> Tentar alterar o tipo de dados do valor geraria uma exceção no tempo de execução? </font>

In [37]:
int first = 2;
string second = "4";
int result = first + second;
Console.WriteLine(result);

- Concatenar um número a um string e salvá-lo em uma variável de cadeia de caracteres:

In [18]:
int first = 2;
string second = "4";
string result = first + second;
Console.WriteLine(result);

24


### <font color=hotpink> Compiladores fazem conversões seguras </font>


O compilador C# identifica um potencial problema na operação proposta. A variável `second` é do tipo `string`, o que permite que ela receba valores distintos, como `"hello"`. Caso o compilador C# tentasse converter um valor como `"hello"` em um número, isso resultaria em uma exceção em tempo de execução. Para prevenir tal eventualidade, o compilador C# não realiza automaticamente a conversão de `string` para `int`.

Para executar a conversão de dados, você pode usar uma de várias técnicas:

- Usar um método auxiliar no tipo de dados
- Usar um método auxiliar na variável
- Use os métodos da classe `Convert`

### <font color=hotpink> Tentar alterar o tipo de dados do valor resultaria em uma perda de informações? </font>

In [19]:
int myInt = 3;
Console.WriteLine($"int: {myInt}");

decimal myDecimal = myInt;
Console.WriteLine($"decimal: {myDecimal}");

int: 3
decimal: 3


- *Como qualquer valor int pode se ajustar facilmente dentro de um decimal, o compilador executa a conversão.*

- O termo ***conversão de expansão*** significa que você está tentando converter um valor de um tipo de dados que poderia armazenar menos informações em um tipo de dados que pode armazenar mais informações. Neste caso, um valor armazenado em uma variável do tipo `int` convertida em uma variável do tipo `decimal` não perde informação.

### <font color=hotpink> Executar uma coerção (casting) </font>

- conversão explícita para o tipo de dados de casting definido

In [20]:
decimal myDecimal = 3.14m;
Console.WriteLine(myDecimal);

3,14


In [21]:
int myInt = (int)myDecimal;
Console.WriteLine(myInt);

3


- O termo ***conversão de restrição*** significa que você está tentando converter um valor de um tipo de dados que pode conter mais informações para um tipo de dados que pode conter menos. Nesse caso, você pode perder informações como precisão (ou seja, o número de valores após o ponto decimal). 

In [22]:
decimal myDecimal = 1.23456789m;
float myFloat = (float)myDecimal;

Console.WriteLine($"Decimal: {myDecimal}");
Console.WriteLine($"Float  : {myFloat}");

Decimal: 1,23456789
Float  : 1,2345679


- O casting de `decimal` em `float` é uma ***conversão de restrição*** porque você está perdendo a precisão.

### <font color=hotpink> Usar `ToString()` para converter um número em um `string` </font>

- Cada variável de tipo de dados tem um método `ToString()`.
- O que o método `ToString()` faz depende de como ele é implementado em um determinado tipo. 

In [23]:
int primeiro = 5;
int segundo = 10;
string mensagem = primeiro.ToString() + segundo.ToString();
Console.WriteLine(mensagem);

510


### <font color=hotpink> Converter um `string` em um `int` usando o método auxiliar `Parse()` </font>

- A maioria dos tipos de dados numéricos tem um método `Parse()`, que converte uma cadeia de caracteres no tipo de dado especificado.

In [24]:
string primeiro = "5";
string segundo = "10";
 
 int soma = int.Parse(primeiro) + int.Parse(segundo);
 Console.WriteLine(soma);


15


### <font color=hotpink> Converter um `string` em um `int` usando a classe `Convert` </font>

- A classe `Convert` tem vários métodos auxiliares para converter um valor de um tipo em outro.

In [25]:
string valor1 = "7";
string valor2 = "11";

int resultadoMultiplicacao = Convert.ToInt32(valor1) * Convert.ToInt32(valor2);

Console.WriteLine(resultadoMultiplicacao);

77


- *O método `ToInt32()` tem 19 sobrecargas, o que permite aceitar praticamente todos os tipos de dados.*

- *A classe `Convert` é melhor para converter números fracionários em números inteiros (`int`) porque os arredonda da maneira esperada.*

### <font color=hotpink> Comparar o casting e a conversão de um `decimal` em um `int` </font>

In [26]:
int valor1 = (int)1.5m; // Trunca o valor na parte inteira.
Console.WriteLine(valor1);

1


In [27]:
int valor2 = Convert.ToInt32(1.5m); // arredonda o valor para inteiro.
Console.WriteLine(valor2);

2


## <font color=mediumorchid> Método TryParse()
___
</font>

Às vezes, ao trabalhar com dados, é necessário converter dados de cadeia de caracteres em um tipo de dados numérico. Para evitar uma exceção de formato, use o método `TryParse()` no tipo de dados de destino.

### <font color=hotpink> Usar `TryParse()` </font>

O método `TryParse()` faz várias coisas simultaneamente:

- Ele tenta analisar uma cadeia de caracteres sobre o tipo de dados numérico fornecido.
- Se a conversão der certo, o valor convertido é armazenado em um parâmetro de saída, explicado na seção a seguir.
- Retorna um `bool` para indicar se a ação foi bem-sucedida ou falhou.

É possível usar o valor de retorno booliano para tomar uma ação com relação ao valor

### <font color=hotpink> Parâmetros externos </font>

- Os métodos podem retornar um valor ou retornar "nulo", o que significa que eles não retornam nenhum valor. 
- Os métodos também podem retornar valores por meio de parâmetros `out`, definidos como um parâmetro de entrada, porém incluem a palavra-chave `out`.

### <font color=hotpink> Utilizar o método TryParse() para analisar uma cadeia de caracteres como um int </font>

In [28]:
string valor = "102";

int resultado = 0;

if (int.TryParse(valor, out resultado))
{
    Console.WriteLine(resultado);
}
else
{
    Console.WriteLine("Não é um número");
}

102


In [29]:
string valor = "opa";

int resultado = 0;

if (int.TryParse(valor, out resultado))
{
    Console.WriteLine(resultado);
}
else
{
    Console.WriteLine("Não é um número");
}

Não é um número


- *A palavra-chave out indica ao compilador que o método TryParse() não retornará um valor somente da maneira tradicional (como um valor de retorno), mas também comunicará uma saída por meio desse parâmetro bidirecional.*

### <font color=hotpink> O int analisado será usado posteriormente no código </font>

In [31]:
string valor = "102";

int resultado = 0;

if (int.TryParse(valor, out resultado))
{
    Console.WriteLine($"Medida: {resultado}");
}
else
{
    Console.WriteLine("Não é um número");
}

Console.WriteLine($"Medida (w/ offset): {50 + resultado}");

Medida: 102
Medida (w/ offset): 152


In [36]:
string valor = "badf";

int resultado = 0;

if (int.TryParse(valor, out resultado))
{
    Console.WriteLine($"Medida: {resultado}");
}
else
{
    Console.WriteLine($"Não foi possível converter o valor \"{valor}\" para inteiro.");
}
if (resultado > 0)
    Console.WriteLine($"Medida (w/ offset): {50 + resultado}");

Não foi possível converter o valor "badf" para inteiro.


# <font color=turquoise> Executar operações em matrizes usando métodos auxiliares em C#
___

## <font color=mediumorchid> Conhecer as funções Sort() e Reverse()
___
</font>

A classe Array contém métodos que você pode usar para manipular o conteúdo, a disposição e o tamanho de uma matriz.

### <font color=hotpink> Criar uma matriz de paletes, depois classificá-los
 </font>

In [44]:
string[] pallets = { "B14", "A11", "B12", "A13" };

Console.WriteLine("Na ordem original:");
foreach (var pallet in pallets) 
{
    Console.Write($"{pallet}\t");
}

Console.WriteLine("");
Console.WriteLine("Ordenados...:");
Array.Sort(pallets);

foreach (var pallet in pallets)
{
    Console.Write($"{pallet}\t");
}

Na ordem original:
B14	A11	B12	A13	
Ordenados...:
A11	A13	B12	B14	

### <font color=hotpink> Inverter a ordem dos paletes </font>

In [46]:
string[] pallets = { "B14", "A11", "B12", "A13" };

Console.WriteLine("Na ordem original:");
foreach (var pallet in pallets) 
{
    Console.Write($"{pallet}\t");
}

Console.WriteLine("\n");
Console.WriteLine("Ordenados...:");
Array.Sort(pallets);

foreach (var pallet in pallets)
{
    Console.Write($"{pallet}\t");
}

Console.WriteLine("\n");
Console.WriteLine("Revertendo a ordem...:");
Array.Reverse(pallets);

foreach (var pallet in pallets)
{
    Console.Write($"{pallet}\t");
}

Na ordem original:
B14	A11	B12	A13	

Ordenados...:
A11	A13	B12	B14	

Revertendo a ordem...:
B14	B12	A13	A11	

## <font color=mediumorchid> Métodos Clear() e Resize()
___
</font>

### <font color=hotpink> Usar métodos de matriz para limpar e redimensionar uma matriz </font>

- O método `Array.Clear()` permite remover o conteúdo de elementos específicos em sua matriz e substituí-lo pelo valor padrão da matriz. 

- O método `Array.Resize()` adiciona ou remove elementos da matriz.

In [52]:
string[] pallets = { "B14", "A11", "B12", "A13" };

Console.WriteLine("Matriz Original:");
foreach (string pallet in pallets) Console.Write($"{pallet}\t");
Console.WriteLine("\n---------------------------------------------------\n");

Array.Clear(pallets, 0, 2); // limpar os valores começando no índice 0 e limpando os 2 elementos.
Console.WriteLine($"Limpando 2 ... count: {pallets.Length}");
foreach (string pallet in pallets) Console.Write($"{pallet}\t");
Console.WriteLine("\n---------------------------------------------------\n");



Matriz Original:
B14	A11	B12	A13	
---------------------------------------------------

Limpando 2 ... count: 4
		B12	A13	
---------------------------------------------------



### <font color=hotpink> Cadeia de caracteres vazia versus nula </font>

- Quando você usa `Array.Clear()`, os elementos que foram limpos não referenciam mais uma cadeia de caracteres na memória. 

> - Acessar o valor de um elemento que foi limpo:

In [53]:
string[] pallets = { "B14", "A11", "B12", "A13" };

Console.WriteLine("Matriz Original:");
foreach (string pallet in pallets) Console.Write($"{pallet}\t");
Console.WriteLine("\n---------------------------------------------------\n");

Console.WriteLine($"Antes: {pallets[0]}");
Array.Clear(pallets, 0, 2); // limpar os valores começando no índice 0 e limpando os 2 elementos.
Console.WriteLine($"Depois: {pallets[0]}");

Matriz Original:
B14	A11	B12	A13	
---------------------------------------------------

Antes: B14
Depois: 


- *Se você se concentrar na linha de saída After: , poderá achar que o valor armazenado em pallets[0] é uma cadeia de caracteres vazia. No entanto, o compilador C# converte implicitamente o valor nulo em uma cadeia de caracteres vazia para apresentação.*

### <font color=hotpink> Chamar um método auxiliar de cadeia de caracteres em um elemento limpo </font>

In [55]:
string[] pallets = { "B14", "A11", "B12", "A13" };

Console.WriteLine("Matriz Original:");
foreach (string pallet in pallets) Console.Write($"{pallet}\t");
Console.WriteLine("\n---------------------------------------------------\n");

Console.WriteLine($"Antes: {pallets[0].ToLower()}");
Array.Clear(pallets, 0, 2); // limpar os valores começando no índice 0 e limpando os 2 elementos.
// Console.WriteLine($"Depois: {pallets[0].ToLower()}");
if (pallets[0] != null)
    Console.WriteLine($"Depois: {pallets[0].ToLower()}");

Matriz Original:
B14	A11	B12	A13	
---------------------------------------------------

Antes: b14


*Essa exceção é gerada porque a tentativa de chamar o método no conteúdo do elemento pallets[0] acontece antes que o compilador do C# tenha a chance de converter implicitamente o valor nulo para uma cadeia de caracteres vazia.*

### <font color=hotpink> Redimensionar a matriz para adicionar mais elementos </font>

- Utiliza o Método `Resize()` passando por referência a matriz `ref matriz`

In [65]:
string[] pallets = { "B14", "A11", "B12", "A13" };
Console.WriteLine($"Total de pallets: {pallets.Length}");
foreach (string pallet in pallets) Console.Write($"- {pallet}\t");
Console.WriteLine("\n--------------------------------------------\n\n");

Console.WriteLine("Redimensionando...");
Array.Resize(ref pallets, 6);
Console.WriteLine($"Total de pallets: {pallets.Length}");
foreach (string pallet in pallets) Console.Write($"- {pallet}\t");
Console.WriteLine("\n--------------------------------------------\n\n");


Total de pallets: 4
- B14	- A11	- B12	- A13	
--------------------------------------------


Redimensionando...
Total de pallets: 6
- B14	- A11	- B12	- A13	- 	- 	
--------------------------------------------




### <font color=hotpink> Redimensionar a matriz para remover elementos </font>

In [72]:
string[] pallets = { "B14", "A11", "B12", "A13" };
Console.WriteLine($"Total de pallets: {pallets.Length}");
foreach (string pallet in pallets) Console.WriteLine($"- {pallet}");
Console.WriteLine("--------------------------------------------\n");

Array.Clear(pallets, 0, 2);
Console.WriteLine($"Clearing 2 ... Total de pallets: {pallets.Length}");
foreach (string pallet in pallets) Console.WriteLine($"- {pallet}");
Console.WriteLine("--------------------------------------------\n");

Array.Resize(ref pallets, 6);
Console.WriteLine($"Resizing...6 Total de pallets: {pallets.Length}");
foreach (string pallet in pallets) Console.WriteLine($"- {pallet}");
Console.WriteLine("--------------------------------------------\n");

pallets[4] = "C01";
pallets[5] = "C02";
Console.WriteLine($"New elements 5 e 6.. Total de pallets: {pallets.Length}");
foreach (string pallet in pallets) Console.WriteLine($"- {pallet}");
Console.WriteLine("--------------------------------------------\n");


Array.Resize(ref pallets, 3); // remove os 3 últimos elementos
Console.WriteLine($"Resizing 3... Total de pallets: {pallets.Length}");
foreach (string pallet in pallets) Console.WriteLine($"- {pallet}\t");
Console.WriteLine("--------------------------------------------\n");

Total de pallets: 4
- B14
- A11
- B12
- A13
--------------------------------------------

Clearing 2 ... Total de pallets: 4
- 
- 
- B12
- A13
--------------------------------------------

Resizing...6 Total de pallets: 6
- 
- 
- B12
- A13
- 
- 
--------------------------------------------

New elements 5 e 6.. Total de pallets: 6
- 
- 
- B12
- A13
- C01
- C02
--------------------------------------------

Resizing 3... Total de pallets: 3
- 	
- 	
- B12	
--------------------------------------------



### <font color=hotpink> É possível remover elementos nulos de uma matriz? </font>

- O método `Array.Resize()` não remove elementos vazios de uma matriz, e não há algum outro método auxiliar que faz isso automaticamente.
    - A melhor maneira de esvaziar os elementos de uma matriz é contar o número de elementos não nulos percorrendo cada item e incrementando uma variável (contador). 
    - Em seguida, você criará uma segunda matriz que tem o tamanho da variável do contador. 
    - Por fim, você executará um loop em cada elemento na matriz original e copiará valores não nulos na nova matriz.

## <font color=mediumorchid> Split() e Join()
___
</font>

### <font color=hotpink> Métodos de matriz do tipo de dados de cadeia de caracteres </font>

As variáveis do tipo string têm muitos métodos internos que convertem uma única cadeia de caracteres em uma matriz de cadeias de caracteres menores ou em uma matriz de caracteres individuais.

### <font color=hotpink> Usar o `ToCharArray()` para reverter um `string` </font>

In [75]:
string valor = "abc123";
char[] valorArray = valor.ToCharArray();

foreach (var i in valorArray) Console.WriteLine($"-- {i}");

-- a
-- b
-- c
-- 1
-- 2
-- 3


### <font color=hotpink> Inverter e, em seguida, combinar a matriz char em uma nova cadeia de caracteres </font>

In [78]:
string valor = "abc123";
char[] valorArray = valor.ToCharArray();
foreach (var i in valorArray) Console.Write($"{i}");
Console.WriteLine("\n-----------------\n");

Array.Reverse(valorArray);
string valor2 = new String(valorArray);
Console.WriteLine($"{valor2}\t");


abc123
-----------------
321cba	


*A expressão new string(valueArray) cria uma instância vazia da classe System.String (que é a mesma do tipo de dados string em C#) e passa a matriz char como um construtor.*

### <font color=hotpink> Combinar todos os caracteres em uma nova cadeia de caracteres com valor separado por vírgula usando `Join()` </font>

In [79]:
string valor = "abc123";
char[] valorArray = valor.ToCharArray();
Array.Reverse(valorArray);

string resultado = String.Join(",", valorArray);

Console.WriteLine(resultado);

3,2,1,c,b,a


### <font color=hotpink> `Split()` a nova cadeia de caracteres de valor separado por vírgula em uma matriz de cadeias de caracteres </font>

In [83]:
string valor = "abc123";
char[] valorArray = valor.ToCharArray();
Array.Reverse(valorArray);

string resultado = String.Join(",", valorArray);
Console.WriteLine($"String: {resultado}");


Console.WriteLine("");
string[] items = resultado.Split(",");
foreach(var item in items) Console.WriteLine(item);

String: (3,2,1,c,b,a)

3
2
1
c
b
a


# <font color=turquoise> Formatar dados alfanuméricos para apresentação em C#
___

## <font color=mediumorchid> Noções básicas da formatação de cadeia de caracteres
___
</font>

### <font color=hotpink> O que é a Formatação Composta? </font>

- A formatação composta usa espaços reservados numerados dentro de uma cadeia de caracteres. 


>___
>
> Este exemplo de formatação composta usa um método interno Format() na palavra-chave do tipo de dados string: 
>
>___

In [85]:
string first = "Hello";
string second = "World";
string result = string.Format("{0} {1}!", first, second);

Console.WriteLine(result);

Hello World!


- Os tipos de dados e as variáveis de um determinado tipo de dados têm ***“métodos auxiliares”*** internos para facilitar determinadas tarefas.
- A cadeia de caracteres literal `"{0} {1}!"` forma um modelo que tem suas partes substituídas no runtime.
- O token `{0}` é substituído pelo primeiro argumento depois do modelo da cadeia de caracteres; em outras palavras, o valor da variável `first`.
- O token `{1}` é substituído pelo segundo argumento depois do modelo da cadeia de caracteres; em outras palavras, o valor da variável `second`.

In [86]:
string first = "Hello";
string second = "World";
Console.WriteLine("{1} {0}!", first, second);
Console.WriteLine("{0} {0} {0}!", first, second);

World Hello!
Hello Hello Hello!


### <font color=hotpink> O que é interpolação de cadeia de caracteres? </font>

- Técnica que simplifica a formatação composta onde você usa apenas o nome da variável dentro das chaves.
- A string deve começar com `$` antes das aspas;


In [87]:
string first = "Hello";
string second = "World";
Console.WriteLine($"{first} {second}!");
Console.WriteLine($"{second} {first}!");
Console.WriteLine($"{first} {first} {first}!");

Hello World!
World Hello!
Hello Hello Hello!


### <font color=hotpink> Formatação de Moeda </font>

- A formatação composta e a interpolação de cadeia de caracteres podem ser usadas para formatar valores para exibição considerando uma linguagem e cultura específicas. 
- O  especificador de formato de moeda :C é usado para apresentar as variáveis como moeda.

In [91]:
decimal preco = 123.45m;
int desconto = 50;

Console.WriteLine($"Preço: {preco:C} (Desconto: {desconto:C})");

Preço: R$ 123,45 (Desconto: R$ 50,00)


*Observe como a adição do :C aos tokens dentro das chaves formata o número como moeda, independentemente se você usa int ou decimal.*

### <font color=hotpink> Como o país/região e o idioma do usuário afetam a formatação da cadeia de caracteres </font>

 - O recurso de formatação de moeda da cadeia de caracteres depende da configuração do computador local para `cultura`. 
    - Nesse contexto, o termo “cultura” refere-se ao país/região e ao idioma do usuário final.
- Por exemplo:
   - o código de cultura de um falante de inglês nos EUA é `en-US`.
   - o código de cultura de um falante de francês na França é `fr-FR`.
   - o código de cultura de um falante de francês no Canadá é `fr-CA`.


> Exemplo US:

In [93]:
using System.Globalization;
CultureInfo.CurrentCulture = new CultureInfo("en-US");

decimal preco = 123.45m;
int desconto = 50;

Console.WriteLine($"Preço: {preco:C} (Desconto: {desconto:C})");

Preço: $123.45 (Desconto: $50.00)


> Exemplo FR:

In [94]:
using System.Globalization;
CultureInfo.CurrentCulture = new CultureInfo("fr-FR");

decimal preco = 123.45m;
int desconto = 50;

Console.WriteLine($"Preço: {preco:C} (Desconto: {desconto:C})");

Preço: 123,45 € (Desconto: 50,00 €)


> Exemplo CN:

In [95]:
using System.Globalization;
CultureInfo.CurrentCulture = new CultureInfo("fr-CN");

decimal preco = 123.45m;
int desconto = 50;

Console.WriteLine($"Preço: {preco:C} (Desconto: {desconto:C})");

Preço: 123,45 CNY (Desconto: 50,00 CNY)


### <font color=hotpink> Formatar números </font>

- O especificador de formato numérico N torna os números mais legíveis.
    - Separa unidade, dezena, centena e milhar...
    - Por padrão, o especificador de formato numérico N exibe apenas dois dígitos após o ponto decimal.
    - Se desejar mostrar mais precisão, adicione um número após o especificador.

In [103]:
decimal medida = 123456.78912m;

Console.WriteLine($"Medida sem formatação: \t{medida} ");
Console.WriteLine($"Medida com formatação: \t{medida:N} ")

Medida sem formatação: 	123456,78912 
Medida com formatação: 	123.456,79 


In [104]:
decimal medida = 123456.78912m;
Console.WriteLine($"Medida com 4 casas após a virgula: \t{medida:N4}");

Medida com 4 casas após a virgula: 	123.456,7891


### <font color=hotpink> csharp </font>

### <font color=hotpink> Formatar percentuais </font>

- Utiliza-se o especificador de formato `P` para formatar percentuais.
- Adiciona-se um número posteriormente para controlar o número de valores exibidos após o ponto decimal. 

In [106]:
decimal taxa = .36785m;
Console.WriteLine($"Taxa: \t{taxa:P}");

Taxa: 	36,79%


In [108]:
decimal taxa = .36785m;
Console.WriteLine($"Taxa: \t{taxa:P2}");

Taxa: 	36,79%


In [109]:
decimal taxa = .36785m;
Console.WriteLine($"Taxa: \t{taxa:P4}");

Taxa: 	36,7850%


### <font color=hotpink> Combinar abordagens de formatação </font>

- Variáveis de cadeia de caracteres podem armazenar as cadeias de caracteres criadas usando técnicas de formatação.

In [117]:
decimal preco = 67.55m;
decimal precoNaPromocao = 59.99m;

string desconto = String.Format("Você economizou na promoção {0:C2} do valor original {1:C2}.", (preco - precoNaPromocao), preco);

Console.WriteLine(desconto);

Você economizou na promoção R$ 7,56 do valor original R$ 67,55.


In [123]:
decimal preco = 67.55m;
decimal precoNaPromocao = 59.99m;

string desconto = String.Format("Você economizou na promoção {0:C2} do valor original {1:C2}.", (preco - precoNaPromocao), preco);


desconto += $" Um desconto de {(preco - precoNaPromocao)/preco:P}!";
Console.WriteLine(desconto)

Você economizou na promoção R$ 7,56 do valor original R$ 67,55. Um desconto de 11,19%!


## <font color=mediumorchid> Interpolação de cadeia de caracteres
___
</font>

Exemplo: Você precisa criar o código para imprimir um recibo para o cliente que compra ações de um produto de investimento. As ações são compradas automaticamente no final do ano com base em uma série de deduções de folha de pagamento, de modo que o número de ações compradas geralmente contém um valor decimal. Para imprimir o recibo, você provavelmente precisará combinar de maneira precisa dados de diferentes tipos, incluindo valores fracionários, moeda e porcentagens.

 
>___
>
> - <font color=hotpink>Exibir o número da fatura usando interpolação de cadeia de caracteres: </font>
>
>___

In [133]:
int numeroDaNota = 1201;
decimal acoesDoProduto = 25.4567m;
decimal valorSubtotal = 2750.00m;
decimal percentualDeImposto = 0.15825m;
decimal valorTotal = 3185.19m;

Console.WriteLine($"Número da Nota: {numeroDaNota}");

Número da Nota: 1201


 
>___
>
> - <font color=hotpink> Exibir as ações de produto com precisão de um milésimo (0,001): </font>
>
>___

In [135]:
int numeroDaNota = 1201;
decimal acoesDoProduto = 25.4567m;
decimal valorSubtotal = 2750.00m;
decimal percentualDeImposto = 0.15825m;
decimal valorTotal = 3185.19m;

Console.WriteLine($"Número da Nota: {numeroDaNota}");
Console.WriteLine($" Ações: {acoesDoProduto:N3} Do Produto");

Número da Nota: 1201
 Ações: 25,457 Do Produto


 
>___
>
> - <font color=hotpink> Exibir o subtotal que você vai cobrar do cliente, formatado como moeda: </font>
>
>___

In [138]:
int numeroDaNota = 1201;
decimal acoesDoProduto = 25.4567m;
decimal valorSubtotal = 2750.00m;
decimal percentualDeImposto = 0.15825m;
decimal valorTotal = 3185.19m;

Console.WriteLine($"Número da Nota: {numeroDaNota}");
Console.WriteLine($" Ações: {acoesDoProduto:N3} Do Produto");
Console.WriteLine($"\t Subtotal: \t{valorSubtotal:C}");

Número da Nota: 1201
 Ações: 25,457 Do Produto
	 Subtotal: 	R$ 2.750,00


 
>___
>
> - <font color=hotpink> Exibir o imposto cobrado na venda, formatado como percentual: </font>
>
>___

In [144]:
int numeroDaNota = 1201;
decimal acoesDoProduto = 25.4567m;
decimal valorSubtotal = 2750.00m;
decimal percentualDeImposto = 0.15825m;
decimal valorTotal = 3185.19m;

Console.WriteLine($"Número da Nota: {numeroDaNota}");
Console.WriteLine($" Ações: {acoesDoProduto:N3} Do Produto");
Console.WriteLine($"\t Subtotal: \t{valorSubtotal:C}");
Console.WriteLine($"\t  Imposto: \t{percentualDeImposto:P}");

Número da Nota: 1201
 Ações: 25,457 Do Produto
	 Subtotal: 	R$ 2.750,00
	  Imposto: 	15,83%


 
>___
>
> - <font color=hotpink> Finalizar o recibo com o valor total devido, formatado como moeda: </font>
>
>___

In [146]:
int numeroDaNota = 1201;
decimal acoesDoProduto = 25.4567m;
decimal valorSubtotal = 2750.00m;
decimal percentualDeImposto = 0.15825m;
decimal valorTotal = 3185.19m;

Console.WriteLine($"Número da Nota: {numeroDaNota}");
Console.WriteLine($"Ações: {acoesDoProduto:N3} Do Produto");
Console.WriteLine($"\t Subtotal: \t{valorSubtotal:C}");
Console.WriteLine($"\t  Imposto: \t{percentualDeImposto:P}");
Console.WriteLine($"\t    Total: \t{valorTotal:C}");

Número da Nota: 1201
Ações: 25,457 Do Produto
	 Subtotal: 	R$ 2.750,00
	  Imposto: 	15,83%
	    Total: 	R$ 3.185,19


## <font color=mediumorchid> Preenchimento e alinhamento
___
</font>

### <font color=hotpink> Métodos internos </font>

O método string.Format() é usado para executar a formatação composta, como no exemplo:

In [147]:
string first = "Hello";
string second = "World";
string result = string.Format("{0} {1}!", first, second);
Console.WriteLine(result);

Hello World!


- uma palavra-chave que representa um tipo de dados tenha métodos que você pode chamar da mesma maneira como você faz na classe Console. 
- Lista de categorias desses métodos internos:
    - Métodos que adicionam espaços em branco para fins de formatação (`PadLeft()`, `PadRight()`);
    - Métodos que comparam duas cadeias de caracteres ou facilitam a comparação (`Trim()`, `TrimStart()`, `TrimEnd()`, `GetHashcode()`, a propriedade `Length`);
    - Métodos que ajudam a determinar o que há dentro de uma cadeia de caracteres ou até mesmo recuperar apenas uma parte da cadeia de caracteres (`Contains()`, `StartsWith()`, `EndsWith()`, `Substring()`);
    - Métodos que alteram o conteúdo da cadeia de caracteres substituindo, inserindo ou removendo partes (`Replace()`, `Insert()`, `Remove()`);
    - Métodos que transformam uma cadeia de caracteres em uma matriz de cadeias de caracteres ou de caracteres (`Split()`, `ToCharArray()`).

### <font color=hotpink> Formatar cadeias de caracteres adicionando um espaço em branco antes ou depois </font>

- O método `PadLeft()` adiciona espaços em branco ao lado esquerdo da cadeia de caracteres

In [148]:
string entrada = "Pad this";
Console.WriteLine(entrada.PadLeft(12));

    Pad this


- Para adicionar espaço ou caracteres ao lado direito da cadeia de caracteres, use o método `PadRight()`.

In [ ]:
string entrada = "Pad this";
Console.WriteLine(entrada.PadRight(12) + "FIM");

Pad this    FIM


### <font color=hotpink> Sobrecarga do Pad() </font>

- Passar qualquer caractere como parâmetro;

In [150]:
string input = "Pad this";

Console.WriteLine(input.PadLeft(12, '-'));

Console.WriteLine(input.PadRight(12, '-'));

----Pad this
Pad this----


### <font color=hotpink> Trabalhar com cadeias de caracteres preenchidas </font>

- Vamos supor que você trabalha para uma empresa de processamento de pagamentos que ainda dá suporte a sistemas de mainframe herdados. Geralmente, esses sistemas exigem que os dados sejam inseridos em colunas específicas. Por exemplo, armazene a ID do pagamento nas colunas 1 a 6, o nome do favorecido nas colunas 7 a 30 e o valor do pagamento nas colunas 31 a 40. Além disso, é importante que o Valor do Pagamento esteja alinhado à direita.

- Foi solicitado que você compilasse um aplicativo que converterá dados do RDBMS no formato de arquivo herdado. Para que a integração funcione corretamente, a primeira etapa é confirmar o formato de arquivo fornecendo aos mantenedores do sistema herdado um exemplo da saída. Posteriormente, você vai criar esse trabalho para enviar centenas ou milhares de pagamentos a serem processados por meio de um arquivo de texto ASCII.

>___
>
> <font color=hotpink> ***Adicionar a ID de Pagamento à saída:*** </font>
>
>___

In [151]:
string pagamentoId = "769C";
var linhaFormatada = pagamentoId.PadRight(6);

Console.WriteLine(linhaFormatada);

769C  


In [152]:
string pagamentoId = "769C";
string nomeCliente = "Mr. Stephen Ortega";

var linhaFormatada = pagamentoId.PadRight(6);
linhaFormatada += nomeCliente.PadRight(24);

Console.WriteLine(linhaFormatada);

769C  Mr. Stephen Ortega      


>___
>
> <font color=hotpink> ***Adicionar o nome do favorecido à saída:*** </font>
>
>___

In [ ]:
string pagamentoId = "769C";
string nomeCliente = "Mr. Stephen Ortega";

var linhaFormatada = pagamentoId.PadRight(6);
linhaFormatada += nomeCliente.PadRight(24);

Console.WriteLine(linhaFormatada);

769C  Mr. Stephen Ortega      


>___
>
> <font color=hotpink> ***Adicionar o valor do pagamento à saída:*** </font>
>
>___

In [153]:
string pagamentoId = "769C";
string nomeCliente = "Mr. Stephen Ortega";
string valorPagamento  = "$5,000.00";


var linhaFormatada = pagamentoId.PadRight(6);
linhaFormatada += nomeCliente.PadRight(24);
linhaFormatada += valorPagamento.PadLeft(10);

Console.WriteLine(linhaFormatada);

769C  Mr. Stephen Ortega       $5,000.00


In [155]:
string pagamentoId = "769C";
string nomeCliente = "Mr. Stephen Ortega";
string valorPagamento  = "$5,000.00";


var linhaFormatada = pagamentoId.PadRight(6);
linhaFormatada += nomeCliente.PadRight(24);
linhaFormatada += valorPagamento.PadLeft(10);
Console.WriteLine("1234567890123456789012345678901234567890");
Console.WriteLine(linhaFormatada);

1234567890123456789012345678901234567890
769C  Mr. Stephen Ortega       $5,000.00


# <font color=turquoise> Modificar o conteúdo de cadeias de caracteres usando métodos internos de tipo de dados em C#
___


## <font color=mediumorchid> Métodos auxiliares IndexOf() e Substring() da cadeia de caracteres
___
</font>

### <font color=hotpink> IndexOf() e Substring() </font>

- Use o método `IndexOf()` e para localizar a posição de um ou mais caracteres ou uma cadeia de caracteres dentro de uma cadeia de caracteres maior.

- Use o método `Substring()` para retornar a parte da cadeia de caracteres maior que segue as posições de caractere especificadas.

### <font color=hotpink> Escreva um código para encontrar pares de parênteses inseridas em uma cadeia de caracteres </font>

In [159]:
string mensagem  = "Find what is (inside the parentheses)";
Console.WriteLine(mensagem);

int indiceAbertura = mensagem.IndexOf('(');
int indiceFechamento = mensagem.IndexOf(')');

Console.WriteLine(indiceAbertura);
Console.WriteLine(indiceFechamento);

Find what is (inside the parentheses)
13
36


## <font color=mediumorchid> 
___
</font>

### <font color=hotpink> Recuperar o valor entre parênteses </font>

In [163]:
string mensagem  = "Find what is (inside the parentheses)";


int indiceAbertura = mensagem.IndexOf('(');
int indiceFechamento = mensagem.IndexOf(')');

int tamanhoTextoParenteses = indiceFechamento - indiceAbertura;
Console.WriteLine(tamanhoTextoPatenteses);


string textoParenteses = mensagem.Substring(indiceAbertura, tamanhoTextoParenteses);

Console.WriteLine(textoParenteses);

22
(inside the parentheses


- ***O método `Substring()` precisa da posição inicial e do número de caracteres, ou comprimento, para recuperar.***

### <font color=hotpink> Modificar a posição inicial da subcadeia de caracteres </font>

In [164]:
string mensagem  = "Find what is (inside the parentheses)";


int indiceAbertura = mensagem.IndexOf('(');
int indiceFechamento = mensagem.IndexOf(')');

indiceAbertura++;

int tamanhoTextoParenteses = indiceFechamento - indiceAbertura;

string textoParenteses = mensagem.Substring(indiceAbertura, tamanhoTextoParenteses);

Console.WriteLine(textoParenteses);

inside the parentheses


In [172]:
string mensagem  = "What is the value <span>between the tags</span>?";


int indiceAbertura = mensagem.IndexOf("<span>");
int indiceFechamento = mensagem.IndexOf("</span>");

indiceAbertura += 6;

int tamanhoTextoParenteses = indiceFechamento - indiceAbertura;

string textoParenteses = mensagem.Substring(indiceAbertura, tamanhoTextoParenteses);

Console.WriteLine(textoParenteses);

between the tags


### <font color=hotpink> Evitar valores mágicos </font>

- As cadeias de caracteres embutidas em código como "<span>" na lista de códigos anterior são conhecidas como "cadeias de caracteres mágicas", e os valores numéricos embutidos em código como 6 são conhecidos como "números mágicos". Esses valores "Mágicos" não são recomendados por muitos motivos e devem ser evitados sempre que possível.

- o código poderá ser desfeito se você embutir a cadeia de caracteres `"<span>"` várias vezes, no código, mas tiver escrito uma instância dela incorretamente como `"<sapn>"`.

- O compilador não detectará `"<sapn>"` em tempo de compilação porque o valor está em uma cadeia de caracteres. O erro de ortografia provavelmente causará problemas em tempo de execução e, dependendo da complexidade do seu código, poderá ser difícil de rastrear.

- Além disso, se você alterar a cadeia de caracteres `"<span>"` para `"<div>"`, mas esquecer de alterar o número 6, o código produzirá resultados indesejáveis.

In [173]:
string message = "What is the value <span>between the tags</span>?";

const string openSpan = "<span>";
const string closeSpan = "</span>";

int openingPosition = message.IndexOf(openSpan);
int closingPosition = message.IndexOf(closeSpan);

openingPosition += openSpan.Length;
int length = closingPosition - openingPosition;
Console.WriteLine(message.Substring(openingPosition, length));

between the tags


*O código usa uma constante com a palavra-chave `const`. Uma constante permite definir e inicializar uma variável cujo valor nunca pode ser alterado. Em seguida, use essa constante no restante do código sempre que precisar desse valor. Isso garante que o valor seja definido apenas uma vez e que a digitação incorreta da variável const seja detectada pelo compilador.*

## <font color=mediumorchid> Métodos auxiliares IndexOfAny() e LastIndexOf() 
___
</font>

- usa-se o método `IndexOfAny()` para encontrar a primeira localização de qualquer um dos string da matriz selecionada. 
- usa-se `LastIndexOf()` para encontrar o local final de uma cadeia de caracteres dentro de outra cadeia de caracteres.

### <font color=hotpink> Recuperar a última ocorrência de uma subcadeia de caracteres </font>

In [179]:
string mensagem = "(What if) I am (only interested) in the last (set of parentheses)?";

int posicaoAbertura = mensagem.LastIndexOf('(');
posicaoAbertura += 1;

int posicaoFechamento = mensagem.LastIndexOf(')');
int tamanho = posicaoFechamento - posicaoAbertura;
Console.WriteLine(mensagem.Substring(posicaoAbertura, tamanho));

set of parentheses


### <font color=hotpink> Recuperar todas as instâncias de subcadeias de caracteres dentro de parênteses </font>

In [1]:
string mensagem = "(What if) I am (only interested) in the last (set of parentheses)?";

while (true) 
{

    int posicaoAbertura = mensagem.IndexOf('(');

    if (posicaoAbertura == -1)  break;

    posicaoAbertura++;
    int posicaoFechamento = mensagem.IndexOf(')');
    int tamanho = posicaoFechamento - posicaoAbertura;
    
    Console.WriteLine(mensagem.Substring(posicaoAbertura, tamanho));

    mensagem = mensagem.Substring(posicaoFechamento + 1);
}

What if
only interested
set of parentheses


- Quando você usa Substring() sem especificar um parâmetro de entrada de comprimento, ele retornará todos os caracteres após a posição inicial especificada.

### <font color=hotpink> Trabalhar com diferentes tipos de conjuntos de símbolos </font>

- Você atualiza a cadeia de caracteres `message` adicionando tipos diferentes de símbolos, como `[]` colchetes e chaves `{}`. Para pesquisar vários símbolos simultaneamente, use em `.IndexOfAny()`. Pesquise com `.IndexOfAny()` para retornar o índice do primeiro símbolo da matriz `openSymbols` encontrada na cadeia de caracteres `message`.

In [5]:
string mensagem = "Help (find) the {opening symbols}";
Console.WriteLine($"Procurando a mensagem: {mensagem}");

char[] openSymbols = {'[', '(', '{' };

int startPosition = 6;
int openingPosition = mensagem.IndexOfAny(openSymbols);

Console.WriteLine($"Encontrado sem utlilizar a posição inicial: {mensagem.Substring(openingPosition)}");

openingPosition = mensagem.IndexOfAny(openSymbols, startPosition);
Console.WriteLine($"Encontrado utlilizando a posição inicial {startPosition}: {mensagem.Substring(openingPosition)}");


Procurando a mensagem: Help (find) the {opening symbols}
Encontrado sem utlilizar a posição inicial: (find) the {opening symbols}
Encontrado utlilizando a posição inicial 6: {opening symbols}


In [6]:
string message = "(What if) I have [different symbols] but every {open symbol} needs a [matching closing symbol]?";

// The IndexOfAny() helper method requires a char array of characters. 
// You want to look for:

char[] openSymbols = { '[', '{', '(' };

// You'll use a slightly different technique for iterating through 
// the characters in the string. This time, use the closing 
// position of the previous iteration as the starting index for the 
//next open symbol. So, you need to initialize the closingPosition 
// variable to zero:

int closingPosition = 0;

while (true)
{
    int openingPosition = message.IndexOfAny(openSymbols, closingPosition);

    if (openingPosition == -1) break;

    string currentSymbol = message.Substring(openingPosition, 1);

    // Now  find the matching closing symbol
    char matchingSymbol = ' ';

    switch (currentSymbol)
    {
        case "[":
            matchingSymbol = ']';
            break;
        case "{":
            matchingSymbol = '}';
            break;
        case "(":
            matchingSymbol = ')';
            break;
    }

    // To find the closingPosition, use an overload of the IndexOf method to specify 
    // that the search for the matchingSymbol should start at the openingPosition in the string. 

    openingPosition += 1;
    closingPosition = message.IndexOf(matchingSymbol, openingPosition);

    // Finally, use the techniques you've already learned to display the sub-string:

    int length = closingPosition - openingPosition;
    Console.WriteLine(message.Substring(openingPosition, length));
}

What if
different symbols
open symbol
matching closing symbol


## <font color=mediumorchid> Usar os métodos Remove() e Replace()
___
</font>


### <font color=hotpink> Usar o método Remove() </font>


>___
>
> <font color=orchid> ***Exemplo: Remover caracteres em locais específicos de uma cadeia de caracteres*** </font>
>
>___

In [7]:
string data = "12345John Smith          5000  3  ";
string updatedData = data.Remove(5, 20);
Console.WriteLine(updatedData);

123455000  3  


- O método Remove() funciona de forma semelhante ao método Substring(). Defina uma posição inicial e o comprimento para remover esses caracteres da cadeia de caracteres.

### <font color=hotpink> Usar o método Replace() </font>

- Use o método Replace() quando precisar substituir um ou mais caracteres por um caractere diferente (ou nenhum caractere). 

>___
>
> <font color=orchid> ***Exemplo: Remover caracteres, independentemente de onde eles aparecem em uma cadeia de caracteres*** </font>
>
>___

In [1]:
string mensagem = "This--is--ex-amp-le--da-ta";
mensagem = mensagem.Replace("--", " ");
Console.WriteLine(mensagem);
Console.WriteLine("");
mensagem = mensagem.Replace("-", "");
Console.WriteLine(mensagem);

This is ex-amp-le da-ta

This is example data


### <font color=hotpink> Exemplo </font>

In [2]:
const string input = "<div><h2>Widgets &trade;</h2><span>5000</span></div>";

string quantity = "";
string output = "";

// Your work here

// Extract the "quantity"
const string openSpan = "<span>";
const string closeSpan = "</span>";

int quantityStart = input.IndexOf(openSpan) + openSpan.Length; // + length of <span> so index at end of <span> tag
int quantityEnd= input.IndexOf(closeSpan);
int quantityLength = quantityEnd - quantityStart;
quantity = input.Substring(quantityStart, quantityLength);
quantity = $"Quantity: {quantity}";

// Set output to input, replacing the trademark symbol with the registered trademark symbol
const string tradeSymbol = "&trade;";
const string regSymbol = "&reg;";
output = input.Replace(tradeSymbol, regSymbol);

// Remove the opening <div> tag
const string openDiv = "<div>";
int divStart = output.IndexOf(openDiv);
output = output.Remove(divStart, openDiv.Length);

// Remove the closing </div> tag and add "Output:" to the beginning
const string closeDiv = "</div>";
int divCloseStart = output.IndexOf(closeDiv);
output = "Output: " + output.Remove(divCloseStart, closeDiv.Length);

Console.WriteLine(quantity);
Console.WriteLine(output);

Quantity: 5000
Output: <h2>Widgets &reg;</h2><span>5000</span>


### <font color=hotpink> csharp </font>